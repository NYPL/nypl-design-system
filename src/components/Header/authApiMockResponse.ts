// The URL endpoint to fetch patron data.
const patronApiUrl = "https://platform.nypl.org/api/v0.1/auth/patron/tokens/";
const accessToken =
  "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJodHRwczpcL1wvd3d3Lm55cGwub3JnIiwic3ViIjoiNjM2NzAyOCIsImF1ZCI6ImFwcF9sb2dpbiIsImlhdCI6MTQ4MjE3NjQ3MCwiZXhwIjoxNDgyMTgwMDcwLCJhdXRoX3RpbWUiOjE0ODIxNzY0NzAsInNjb3BlIjoib3BlbmlkIG9mZmxpbmVfYWNjZXNzIGNvb2tpZSBwYXRyb246cmVhZCJ9.JO7VbOqCC7HyjRmeyHD4zM1Gl0JBk5RdxjAkCp0h6sfVe-xs5FyY7biYqs19k4dUY2DbFYR5IG3xYt9IdhqyMkSnJxtiCY36WN7X_e0eBF2T1_IWKGaBc4JlbroMj5_aNB5W4nQvclrdlb2mV38Q_HGAMUKe8DDeCmAHctEtqGppNl8DC7IvqkekRS_6zgQwsHHW5kJR-f7zUROi4fvFpdNR-I7J4VNWdFIOijb4vXFOOWRLzdY_GHLJdWvSgxhqzwkceA5BScCicAKeHYHo04vabNp5TvPXoR0ypULqTyGYsNnXnUmh2Mu46j3bcNTACEKS97FBx1IfwttBL1ARtQ";
export const patronApiUrlWithToken = `${patronApiUrl}${accessToken}`;

// The URL endpoint to refresh the access token.
const authServerDomain = "https://login.nypl.org/auth";
export const tokenRefreshLink = `${authServerDomain}/refresh`;

// The following three variables are mock responses
// from the platform.nypl.org API.
export const mockResponseData = {
  data: {
    decodedToken: {
      iss: "https://www.nypl.org",
      sub: "6367028",
      aud: "app_login",
      iat: 1482176470,
      exp: 1482180070,
      // eslint-disable-next-line camelcase
      auth_time: 1482176470,
      scope: "openid offline_access cookie patron:read",
    },
    patron: {
      id: "6367028",
      updatedDate: "2016-12-13T19:49:57+00:00",
      createdDate: "2015-09-03T14:16:20+00:00",
      deletedDate: null,
      deleted: false,
      suppressed: false,
      names: ["PATRON, JANE"],
      barCodes: ["23333097542730"],
      expirationDate: "2018-09-03",
      homeLibraryCode: "ma",
      birthDate: "1972-05-13",
      emails: ["janepatron@nypl.org"],
    },
  },
  count: 1,
  statusCode: 200,
  debugInfo: [],
};

export const mockErrorResponseData = {
  expired: false,
  statusCode: 400,
  type: "exception",
  message: "No response",
  error: {
    type: "NYPL\\Starter\\APIException",
    code: 0,
    message: "No response",
    file: "/var/task/src/Controller/AuthController.php",
    line: 47,
    trace: [
      "#0 /var/task/src/Controller/AuthController.php(122): NYPL\\Services\\Controller\\AuthController->getTokenResponse('eyJ0eXAiOiJKV1Q...')",
      "#1 /var/task/index.php(260): NYPL\\Services\\Controller\\AuthController->getToken('eyJ0eXAiOiJKV1Q...')",
      "#2 [internal function]: Closure->{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#3 /var/task/vendor/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(41): call_user_func(Object(Closure), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#4 /var/task/vendor/slim/slim/Slim/Route.php(325): Slim\\Handlers\\Strategies\\RequestResponse->__invoke(Object(Closure), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#5 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(116): Slim\\Route->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#6 /var/task/vendor/slim/slim/Slim/Route.php(297): Slim\\Route->callMiddlewareStack(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#7 /var/task/vendor/slim/slim/Slim/App.php(443): Slim\\Route->run(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#8 /var/task/vendor/nypl/microservice-starter/src/Service.php(63): Slim\\App->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#9 [internal function]: NYPL\\Starter\\Service->NYPL\\Starter\\{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#10 /var/task/vendor/slim/slim/Slim/DeferredCallable.php(43): call_user_func_array(Object(Closure), Array)",
      "#11 [internal function]: Slim\\DeferredCallable->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#12 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(67): call_user_func(Object(Slim\\DeferredCallable), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#13 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(116): Slim\\App->Slim\\{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#14 /var/task/vendor/slim/slim/Slim/App.php(337): Slim\\App->callMiddlewareStack(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#15 /var/task/vendor/slim/slim/Slim/App.php(298): Slim\\App->process(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#16 /var/task/index.php(333): Slim\\App->run()",
      "#17 {main}",
    ],
  },
  debugInfo: [],
};

export const mockExpiredResponseData = {
  data: {
    expired: true,
    statusCode: 401,
    type: "exception",
    message: "Expired token",
  },
  error: {
    type: "NYPL\\Starter\\APIException",
    code: 0,
    message: "Expired token",
    file: "/var/task/src/Controller/AuthController.php",
    line: 47,
    trace: [
      "#0 /var/task/src/Controller/AuthController.php(122): NYPL\\Services\\Controller\\AuthController->getTokenResponse('eyJ0eXAiOiJKV1Q...')",
      "#1 /var/task/index.php(260): NYPL\\Services\\Controller\\AuthController->getToken('eyJ0eXAiOiJKV1Q...')",
      "#2 [internal function]: Closure->{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#3 /var/task/vendor/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(41): call_user_func(Object(Closure), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#4 /var/task/vendor/slim/slim/Slim/Route.php(325): Slim\\Handlers\\Strategies\\RequestResponse->__invoke(Object(Closure), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Array)",
      "#5 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(116): Slim\\Route->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#6 /var/task/vendor/slim/slim/Slim/Route.php(297): Slim\\Route->callMiddlewareStack(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#7 /var/task/vendor/slim/slim/Slim/App.php(443): Slim\\Route->run(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#8 /var/task/vendor/nypl/microservice-starter/src/Service.php(63): Slim\\App->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#9 [internal function]: NYPL\\Starter\\Service->NYPL\\Starter\\{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#10 /var/task/vendor/slim/slim/Slim/DeferredCallable.php(43): call_user_func_array(Object(Closure), Array)",
      "#11 [internal function]: Slim\\DeferredCallable->__invoke(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#12 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(67): call_user_func(Object(Slim\\DeferredCallable), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(NYPL\\Starter\\Service))",
      "#13 /var/task/vendor/slim/slim/Slim/MiddlewareAwareTrait.php(116): Slim\\App->Slim\\{closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#14 /var/task/vendor/slim/slim/Slim/App.php(337): Slim\\App->callMiddlewareStack(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#15 /var/task/vendor/slim/slim/Slim/App.php(298): Slim\\App->process(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))",
      "#16 /var/task/index.php(333): Slim\\App->run()",
      "#17 {main}",
    ],
  },
  debugInfo: [],
};

export const mockLoginCookie =
  '{"access_token":"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJodHRwczpcL1wvd3d3Lm55cGwub3JnIiwic3ViIjoiNjM2NzAyOCIsImF1ZCI6ImFwcF9sb2dpbiIsImlhdCI6MTQ4MjE3NjQ3MCwiZXhwIjoxNDgyMTgwMDcwLCJhdXRoX3RpbWUiOjE0ODIxNzY0NzAsInNjb3BlIjoib3BlbmlkIG9mZmxpbmVfYWNjZXNzIGNvb2tpZSBwYXRyb246cmVhZCJ9.JO7VbOqCC7HyjRmeyHD4zM1Gl0JBk5RdxjAkCp0h6sfVe-xs5FyY7biYqs19k4dUY2DbFYR5IG3xYt9IdhqyMkSnJxtiCY36WN7X_e0eBF2T1_IWKGaBc4JlbroMj5_aNB5W4nQvclrdlb2mV38Q_HGAMUKe8DDeCmAHctEtqGppNl8DC7IvqkekRS_6zgQwsHHW5kJR-f7zUROi4fvFpdNR-I7J4VNWdFIOijb4vXFOOWRLzdY_GHLJdWvSgxhqzwkceA5BScCicAKeHYHo04vabNp5TvPXoR0ypULqTyGYsNnXnUmh2Mu46j3bcNTACEKS97FBx1IfwttBL1ARtQ","expires_in":3600,"token_type":"Bearer","scope":"openid+offline_access+cookie+patron:read","refresh_token":"02b49603a8a2719389a6c77416b110675067827d"}';
