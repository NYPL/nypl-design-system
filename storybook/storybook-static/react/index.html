<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>