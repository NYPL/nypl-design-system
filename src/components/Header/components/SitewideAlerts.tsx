import { chakra, useStyleConfig } from "@chakra-ui/react";
import React, { useState, useEffect } from "react";

import List from "../../List/List";
import Notification from "../../Notification/Notification";
import { Alert, alertsApiUrl, parseAlertsData } from "../headerUtils";

/**
 * The SitewideAlerts component makes an API request to an NYPL API endpoint
 * to fetch NYPL sitewide alerts. While this component can be used in
 * isolation, it is already rendered in the DS Header component.
 */
export const SitewideAlerts = chakra(({ isMobile }) => {
  const [alerts, setAlerts] = useState<Alert[]>([]);
  const styles = useStyleConfig("SitewideAlerts", { isMobile });
  const fetchErrorMessage =
    "NYPL Reservoir SitewideAlerts: There was an error fetching NYPL sitewide alerts.";

  // Make a request to the NYPL API endpoint for sitewide alerts, parse
  // the data, and set it to the local state.
  useEffect(() => {
    fetch(alertsApiUrl)
      .then((response) => {
        if (response.status >= 200 && response.status < 300) {
          return response.json();
        } else {
          throw new Error(fetchErrorMessage);
        }
      })
      .then((data) => {
        const parsedData: Alert[] = parseAlertsData(data);
        setAlerts(parsedData);
      })
      .catch((_error) => {
        // Override any error message with our own.
        console.warn(fetchErrorMessage);
      });
  }, []);

  // Great, we have NYPL alerts data. Each alert is rendered as a list
  // item in an unordered list. Note that the alert description may
  // contain HTML which is rendered as-is.
  const getAlertsElems = (data: Alert[]) => {
    return (
      <List noStyling type="ul">
        {data.map((alert: Alert) => (
          <li
            key={alert.id}
            dangerouslySetInnerHTML={{ __html: alert.description }}
          />
        ))}
      </List>
    );
  };

  return alerts.length > 0 ? (
    <Notification
      aria-label="Sitewide alerts"
      id="sitewideAlerts-notification"
      isCentered
      noMargin
      notificationContent={getAlertsElems(alerts)}
      showIcon={false}
      __css={styles}
    />
  ) : null;
});

export default SitewideAlerts;
