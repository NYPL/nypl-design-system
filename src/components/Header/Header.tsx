import React, { useState } from "react";
import {
  chakra,
  Box,
  HStack,
  useMultiStyleConfig,
  Spacer,
  VStack,
} from "@chakra-ui/react";

import HorizontalRule from "../HorizontalRule/HorizontalRule";
import Link from "../Link/Link";
import Logo from "../Logo/Logo";
import SkipNavigation from "../SkipNavigation/SkipNavigation";
/** Internal Header-only components */
import UpperNav from "./components/UpperNav";
import LowerNav from "./components/LowerNav";
import Mobile from "./components/Mobile";
import SitewideAlerts from "./components/SitewideAlerts";
import useWindowSize from "../../hooks/useWindowSize";

export const Header = chakra(() => {
  const windowDimensions = useWindowSize();
  // These values are based on the breakpoints md and lg
  // from the NYPL theme object.
  const breakpointMedium = 600;
  const breakpointLarge = 960;

  const isWidthMobile = windowDimensions.width < breakpointMedium;
  const isWidthLarge = windowDimensions.width > breakpointLarge;

  const styles = useMultiStyleConfig("Header", {
    isWidthMobile,
    isWidthLarge,
  });

  const [loginOpen, setLoginOpen] = useState<boolean>(false);

  return (
    <Box __css={styles.container}>
      <SkipNavigation />
      <SitewideAlerts />
      <header>
        <HStack id="headerMain" justifyContent="space-between">
          <Link
            aria-label="The New York Public Library"
            href="https://nypl.org"
            __css={styles.logo}
          >
            <Logo
              id="header-nypl-logo"
              name={isWidthLarge ? "nyplFullBlack" : "nyplLionBlack"}
              size={isWidthLarge ? "small" : "xxsmall"}
              title="NYPL Header Logo"
            />
          </Link>
          <Spacer />
          {!isWidthMobile ? (
            <VStack spacing="65px">
              <UpperNav loginOpen={loginOpen} setLoginOpen={setLoginOpen} />
              <LowerNav />
            </VStack>
          ) : (
            <Mobile />
          )}
        </HStack>
        <HorizontalRule __css={styles.horizontalRule} />
      </header>
    </Box>
  );
});

export default Header;
