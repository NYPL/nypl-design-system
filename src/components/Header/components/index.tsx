import Login from "./Login";
import LoginButton from "./LoginButton";
import LowerNav from "./LowerNav";
import MobileIconNav from "./MobileIconNav";
import MobileNav from "./MobileNav";
import MobileNavButton from "./MobileNavButton";
import SearchButton from "./SearchButton";
import SearchForm from "./SearchForm";
import SitewideAlerts from "./SitewideAlerts";
import UpperNav from "./UpperNav";

export default {
  Login,
  LoginButton,
  LowerNav,
  MobileIconNav,
  MobileNav,
  MobileNavButton,
  SearchButton,
  SearchForm,
  SitewideAlerts,
  UpperNav,
};
