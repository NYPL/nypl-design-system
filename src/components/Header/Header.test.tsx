import * as React from "react";
import Cookies from "js-cookie";
import { render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import Header from "./Header";
import { refineryResponse } from "./components/SitewideAlerts.test";

const patronResponse = {
  data: {
    patron: {
      barCodes: ["01234567890123"],
      emails: ["test@nypl.org"],
      id: "0123456",
      names: ["PATRON, JANE"],
    },
  },
  count: 1,
  debugInfo: [],
  statusCode: 200,
};

const mockCookieValue = '{"token_type":"Bearer", "access_token":"123abc"}';

describe("Header Accessibility", () => {
  it("passes axe accessibility test", async () => {
    // Mock the fetch API call in `SitewideAlerts`.
    (global as any).fetch = jest.fn(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(refineryResponse),
      })
    ) as jest.Mock;

    const { container } = await waitFor(() => render(<Header />));
    expect(await axe(container)).toHaveNoViolations();
  });
});

// TODO: These tests do not currently test the mobile web view.
// We need to determine a way of doing this for all responsive
// components, and will add this in at a later date.
describe("Header", () => {
  beforeEach(async () => {
    // Mock the fetch API call in `SitewideAlerts`.
    (global as any).fetch = jest.fn(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(refineryResponse),
      })
    ) as jest.Mock;

    await waitFor(() => render(<Header />));
  });
  it("renders a skip navigation", () => {
    const skipNavigation = screen.getAllByRole("navigation")[0];
    const { getByRole } = within(skipNavigation);

    expect(skipNavigation).toBeInTheDocument();
    expect(skipNavigation).toHaveAttribute("aria-label", "Skip Navigation");
    expect(getByRole("list")).toBeInTheDocument();
  });
  it("renders a notification", () => {
    const notification = screen.getByRole("complementary");
    const { getByText } = within(notification);

    expect(notification).toBeInTheDocument();
    expect(getByText(/masks are encouraged/i)).toBeInTheDocument();
  });

  it("renders the NYPL logo", () => {
    const { container } = render(<Header />);

    expect(container.querySelectorAll("svg")[0]).toHaveAttribute(
      "title",
      "NYPL Header Logo"
    );
  });

  it("renders the upper links", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The first list is the skip navigation.
    // The second list is the list of alerts in the `SitewideAlerts` component.
    // The third list is the upper navigation.
    const upperList = screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");

    expect(upperLinks.length).toEqual(6);
    expect(upperLinks[0]).toHaveTextContent(/log in/i);
    expect(upperLinks[5]).toHaveTextContent(/shop/i);
  });

  it("renders the lower links", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The first list is the skip navigation.
    // The second list is the list of alerts in the `SitewideAlerts` component.
    // The third list is the upper navigation.
    // The fourth list is the lower navigation.
    const lowerList = screen.getAllByRole("list")[3];
    const lowerLinks = within(lowerList).getAllByRole("listitem");

    expect(lowerLinks.length).toEqual(8);
    expect(lowerLinks[0]).toHaveTextContent(/books/i);
    expect(lowerLinks[7]).toHaveTextContent(/search/i);
  });

  it("opens the login menu", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The third list is the upper navigation links.
    let upperList = screen.getAllByRole("list")[2];
    let upperLinks = within(upperList).getAllByRole("listitem");

    expect(upperLinks.length).toEqual(6);

    const logInButton = within(upperLinks[0]).getByRole("button");
    expect(upperLinks[0]).toHaveTextContent(/log in/i);
    expect(upperLinks[1]).toHaveTextContent(/locations/i);

    userEvent.click(logInButton);

    upperList = screen.getAllByRole("list")[2];
    upperLinks = within(upperList).getAllByRole("listitem");

    // Login menu opens, revealing two additional list items.
    expect(upperLinks.length).toEqual(8);
    expect(upperLinks[0]).toHaveTextContent(/close/i);
    expect(upperLinks[1]).toHaveTextContent(/log into/i);
  });

  it("renders the horizontal rule", () => {
    expect(screen.getByRole("separator")).toBeInTheDocument();
  });

  it("renders the UI snapshot correctly", () => {
    const header = renderer.create(<Header />).toJSON();

    expect(header).toMatchSnapshot();
  });
});

describe("Patron is logged in", () => {
  beforeAll(async () => {
    jest.spyOn(Cookies, "get").mockReturnValue(mockCookieValue);

    // This is assuming that the sitewide alerts fetch call runs
    // before the patron information call. If the alerts call runs
    // after the patron information call, the wrong response will be
    // returned for each and this test will fail.
    (global as any).fetch = jest
      .fn()
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(patronResponse),
        })
      );

    await waitFor(() => render(<Header />));
  });
  it("displays logged in view", async () => {
    document.getElementById("chakra-toast-portal")?.remove();

    // The third list is the upper navigation links.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");
    let greeting = screen.queryByText(/jane/i);
    let logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).not.toBeInTheDocument();
    expect(logoutButton).not.toBeInTheDocument();

    expect(firstListElement).toHaveTextContent(/account/i);

    userEvent.click(firstListElement);

    greeting = screen.queryByText(/jane/i);
    logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).toBeInTheDocument();
    expect(logoutButton).toBeInTheDocument();
  });
});

describe("Patron API call fails", () => {
  beforeAll(async () => {
    jest.spyOn(Cookies, "get").mockReturnValue(mockCookieValue);

    (global as any).fetch = jest
      .fn()
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          status: 500,
          json: () => Promise.reject(),
        })
      );

    await waitFor(() => render(<Header />));
  });
  it("logs warning if call fails", async () => {
    const warn = jest.spyOn(console, "warn");
    expect(warn).toHaveBeenCalledWith(
      "NYPL Reservoir Header: There was an error fetching NYPL patron data."
    );

    document.getElementById("chakra-toast-portal")?.remove();

    // If the call fails, the Header should render the logged out state.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");

    expect(firstListElement).toHaveTextContent(/log in/i);
  });
});

describe("Patron API returns wrong data", () => {
  beforeAll(async () => {
    jest.spyOn(Cookies, "get").mockReturnValue(mockCookieValue);

    (global as any).fetch = jest
      .fn()
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve({}),
        })
      );

    await waitFor(() => render(<Header />));
  });
  it("doesn't show logged in view if wrong data is received", async () => {
    document.getElementById("chakra-toast-portal")?.remove();

    // Without the correct data, the Header should render the logged out state.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");

    expect(firstListElement).toHaveTextContent(/log in/i);
  });
});
