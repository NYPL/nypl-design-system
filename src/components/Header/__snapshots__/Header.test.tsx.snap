// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the UI snapshot correctly 1`] = `
<div
  className="css-0"
>
  <nav
    aria-label="Skip Navigation"
    className="css-1xdhyk6"
  >
    <ul
      className="css-1xdhyk6"
    >
      <li>
        <a
          className="css-1xdhyk6"
          href="#mainContent"
          rel={null}
          target={null}
        >
          Skip to Main Content
        </a>
      </li>
      <li>
        <a
          className="css-1xdhyk6"
          href="https://www.nypl.org/accessibility"
          rel={null}
          target={null}
        >
          Click to learn about accessibility at the Library
        </a>
      </li>
    </ul>
  </nav>
  <header>
    <div
      className="chakra-stack css-1oeb4ru"
      id="headerMain"
    >
      <a
        aria-label="The New York Public Library"
        className="css-1xdhyk6"
        href="https://nypl.org"
        rel={null}
        target={null}
      >
        <svg
          aria-hidden={false}
          className="chakra-icon css-1grhd2q"
          focusable={false}
          id="header-nypl-logo"
          role="img"
          title="NYPL Header Logo"
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            strokeWidth="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              strokeLinecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              strokeLinecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              strokeMiterlimit="10"
            />
          </g>
        </svg>
      </a>
      <div
        className="css-17xejub"
      />
      <div
        className="css-k008qs"
      >
        <div
          className="chakra-stack css-1p1m4ay"
        >
          <svg
            aria-hidden={true}
            className="chakra-icon css-1grhd2q"
            focusable={false}
            id="log-in-icon"
            role="img"
            title="Log in to your account"
            viewBox="0 0 24 24"
          >
            <g
              stroke="currentColor"
              strokeWidth="1.5"
            >
              <path
                d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                fill="none"
                strokeLinecap="round"
              />
              <path
                d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                fill="currentColor"
                strokeLinecap="round"
              />
              <circle
                cx="12"
                cy="12"
                fill="none"
                r="11.25"
                strokeMiterlimit="10"
              />
            </g>
          </svg>
          <svg
            aria-hidden={true}
            className="chakra-icon css-1grhd2q"
            focusable={false}
            id="locator-icon"
            role="img"
            title="NYPL Locator"
            viewBox="0 0 24 24"
          >
            <g
              stroke="currentColor"
              strokeWidth="1.5"
            >
              <path
                d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                fill="none"
                strokeLinecap="round"
              />
              <path
                d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                fill="currentColor"
                strokeLinecap="round"
              />
              <circle
                cx="12"
                cy="12"
                fill="none"
                r="11.25"
                strokeMiterlimit="10"
              />
            </g>
          </svg>
          <svg
            aria-hidden={true}
            className="chakra-icon css-1grhd2q"
            focusable={false}
            id="search-icon"
            role="img"
            title="Search"
            viewBox="0 0 24 24"
          >
            <g
              stroke="currentColor"
              strokeWidth="1.5"
            >
              <path
                d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                fill="none"
                strokeLinecap="round"
              />
              <path
                d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                fill="currentColor"
                strokeLinecap="round"
              />
              <circle
                cx="12"
                cy="12"
                fill="none"
                r="11.25"
                strokeMiterlimit="10"
              />
            </g>
          </svg>
          <svg
            aria-hidden={true}
            className="chakra-icon css-1grhd2q"
            focusable={false}
            id="hamburger-icon"
            role="img"
            title="Menu"
            viewBox="0 0 24 24"
          >
            <g
              stroke="currentColor"
              strokeWidth="1.5"
            >
              <path
                d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
                fill="none"
                strokeLinecap="round"
              />
              <path
                d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
                fill="currentColor"
                strokeLinecap="round"
              />
              <circle
                cx="12"
                cy="12"
                fill="none"
                r="11.25"
                strokeMiterlimit="10"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
    <hr
      className="css-mvtkpx"
    />
  </header>
</div>
`;

exports[`SitewideAlerts Successful API request renders the UI snapshot correctly 1`] = `null`;
